lib/irb/type_completion/types.rb

```diff
@@ -0,0 +1,443 @@
+          method.method_types.map do |method_type|
+            score = 0
+            score += 2 if !!method_type.block == has_block
+            reqs = method_type.type.required_positionals
+            opts = method_type.type.optional_positionals
+            rest = method_type.type.rest_positionals
+            trailings = method_type.type.trailing_positionals
+            keyreqs = method_type.type.required_keywords
+            keyopts = method_type.type.optional_keywords
+            keyrest = method_type.type.rest_keywords
+            args = args_types
+            if kwargs_type&.any? && keyreqs.empty? && keyopts.empty? && keyrest.nil?
+              kw_value_type = UnionType[*kwargs_type.values]
+              args += [InstanceType.new(Hash, K: SYMBOL, V: kw_value_type)]
+            end
+            if has_splat
+              score += 1 if args.count { !(_1.is_a? Splat) } <= reqs.size + opts.size + trailings.size
+            elsif reqs.size + trailings.size <= args.size && (rest || args.size <= reqs.size + opts.size + trailings.size)
+              score += 2
+              centers = args[reqs.size...-trailings.size]
+              given = args.first(reqs.size) + centers.take(opts.size) + args.last(trailings.size)
+              expected = (reqs + opts.take(centers.size) + trailings).map(&:type)
+              if rest
+                given << UnionType[*centers.drop(opts.size)]
+                expected << rest.type
+              end
+              if given.any?
+                score += given.zip(expected).count do |t, e|
+                  e = from_rbs_type e, receiver_type
+                  intersect?(t, e) || (intersect?(STRING, e) && t.methods.include?(:to_str)) || (intersect?(INTEGER, e) && t.methods.include?(:to_int)) || (intersect?(ARRAY, e) && t.methods.include?(:to_ary))
+                end.fdiv(given.size)
+              end
+            end
+            [[method_type, given || [], expected || []], score]
+          end
+        end
+        max_score = methods_with_score.map(&:last).max
+        methods_with_score.select { _2 == max_score }.map(&:first)
+      end
+
+      def self.intersect?(a, b)
+        atypes = a.types.group_by(&:class)
+        btypes = b.types.group_by(&:class)
+        if atypes[SingletonType] && btypes[SingletonType]
+          aa, bb = [atypes, btypes].map {|types| types[SingletonType].map(&:module_or_class) }
+          return true if (aa & bb).any?
+        end
+
+        aa, bb = [atypes, btypes].map {|types| (types[InstanceType] || []).map(&:klass) }
+        (aa.flat_map(&:ancestors) & bb).any?
+      end
+
+      def self.type_from_object(object)
+        case object
+        when Array, Hash, Module
+          type_from_object_recursive(object, max_level: 4)
+        else
+          klass = object.singleton_class rescue object.class
+          InstanceType.new klass
+        end
+      end
+
+      def self.type_from_object_recursive(object, max_level:)
```

Comment:
This method is rather complicated. Can we have some unit tests for it?

I'm also a bit worry about the performance impact of resolving types recursively. If say we have a tree structure that's built with a hash, users would probably only need completion on it for 1 or 2 levels? But then we'd need to resolve the entire tree for it, which could be super expensive.

Do you think we can hold off on this for now, and wait for users' request on it?

Reply:
Nice catch :+1:

Currently, the level is restricted to 4, and max sampling size for each hash/array is 1000, but for very complicated structure (large or recursive), the performance was terrible.
```ruby
irb> a=[];50.times{a<<a}
irb> a. # takes about one second
```

> Do you think we can hold off on this for now, and wait for users' request on it?

Yes, we should hold off. 1 level is enough for the first version.

Possible future choices:
- carefully increase level
- restrict by total count of type object, not by depth
- lazy evaluation

Reply:
done in 12efb40d0edb0078ae2d0a532dd225b54ef6ba25